zkevm-circuits/src/ecc_circuit.rs

```diff
@@ -0,0 +1,688 @@
+                    y_cells: y_cells.clone(),
+                };
+                G1Assigned {
+                    decomposed,
+                    x_rlc: pairing_chip.fp_chip.range.gate.inner_product(
+                        ctx,
+                        x_cells,
+                        powers_of_rand.iter().cloned(),
+                    ),
+                    y_rlc: pairing_chip.fp_chip.range.gate.inner_product(
+                        ctx,
+                        y_cells,
+                        powers_of_rand.iter().cloned(),
+                    ),
+                }
+            })
+            .collect_vec();
+        let g2s = op
+            .inputs
+            .iter()
+            .map(|i| {
+                let [x_c0_cells, x_c1_cells, y_c0_cells, y_c1_cells] = self.decompose_g2(i.1);
+                let decomposed = G2Decomposed {
+                    ec_point: pairing_chip.load_private_g2(ctx, Value::known(i.1)),
+                    x_c0_cells: x_c0_cells.clone(),
+                    x_c1_cells: x_c1_cells.clone(),
+                    y_c0_cells: y_c0_cells.clone(),
+                    y_c1_cells: y_c1_cells.clone(),
+                };
+                G2Assigned {
+                    decomposed,
+                    x_c0_rlc: pairing_chip.fp_chip.range.gate.inner_product(
+                        ctx,
+                        x_c0_cells,
+                        powers_of_rand.iter().cloned(),
+                    ),
+                    x_c1_rlc: pairing_chip.fp_chip.range.gate.inner_product(
+                        ctx,
+                        x_c1_cells,
+                        powers_of_rand.iter().cloned(),
+                    ),
+                    y_c0_rlc: pairing_chip.fp_chip.range.gate.inner_product(
+                        ctx,
+                        y_c0_cells,
+                        powers_of_rand.iter().cloned(),
+                    ),
+                    y_c1_rlc: pairing_chip.fp_chip.range.gate.inner_product(
+                        ctx,
+                        y_c1_cells,
+                        powers_of_rand.iter().cloned(),
+                    ),
+                }
+            })
+            .collect_vec();
+
+        // RLC over the entire input bytes.
+        let input_cells = std::iter::empty()
+            .chain(g1s.iter().map(|g1| g1.decomposed.x_cells.clone()))
+            .chain(g1s.iter().map(|g1| g1.decomposed.y_cells.clone()))
+            .chain(g2s.iter().map(|g2| g2.decomposed.x_c0_cells.clone()))
+            .chain(g2s.iter().map(|g2| g2.decomposed.x_c1_cells.clone()))
+            .chain(g2s.iter().map(|g2| g2.decomposed.y_c0_cells.clone()))
+            .chain(g2s.iter().map(|g2| g2.decomposed.y_c1_cells.clone()))
+            .flatten()
+            .collect::<Vec<QuantumCell<F>>>();
+        let input_rlc = pairing_chip.fp_chip.range.gate.inner_product(
+            ctx,
+            input_cells,
+            powers_of_rand.iter().cloned(),
```

Comment:
each pair of `(g1, g2)` is 96 bytes. we support 4 of those. The `input_cells` maximum length will be `4 * 96 = 384`.
The `powers_of_rand` only has `192` challenges. 

`inner_product_simple_with_assignments` seems to pad the rest `192` cells with `0`s

Reply:
cool. thanks for checking